### PR TITLE
Settling times for the SR-570 Preamplifier

### PR DESCRIPTION
**Pull request by canismarko**
When changing gain on the SR-570 pre-amp, the voltage output spikes momentarily, and then settles down. Lower gain values settle faster. This PR adds a settling time to gain-related signals that depends on the target gain, based on measurements of the pre-amps at sector 25-ID-C. Presumably, other SR-570 preamps should be similar.

Fixes https://github.com/BCDA-APS/apstools/issues/906

### CODE REVIEW

**Comment by canismarko** [comment]
By the way, this PR does add a lot of tests (~72), but they're parameterized and run in under a second. Hopefully that's okay. They basically test all the possible combinations of gain values.

File: apstools/devices/tests/test_srs570_preamplifier.py [hunk 1]
```diff
@@ -0,0 +1,148 @@
+import pytest
+from unittest import mock
+
+from apstools.devices.srs570_preamplifier import SRS570_PreAmplifier, GainSignal, DEFAULT_WRITE_TIMEOUT
```
Comment (prjemian) [inline apstools/devices/tests/test_srs570_preamplifier.py:4]:
```suggestion
from ..srs570_preamplifier import SRS570_PreAmplifier, GainSignal, DEFAULT_WRITE_TIMEOUT
```

File: apstools/devices/srs570_preamplifier.py [hunk 2]
```diff
@@ -21,12 +21,169 @@
...
+            if mode == "LOW DRIFT":
+                mode = "LOW NOISE"
+            # Calculate settling time
+            _settle_time = self._settle_time(gain_value=val, gain_unit=unit, gain_mode=mode)
+        else:
+            _settle_time = settle_time
+        return super().set(value, timeout=timeout, settle_time=_settle_time)
+
+
+class GainSignal(GainMixin, EpicsSignal):
```
Comment (prjemian) [inline apstools/devices/srs570_preamplifier.py:178]: Is it necessary to separate the code of GainMixin from GainSignal?  In addition to the use here, will the mixin be used somewhere else?

**Comment by prjemian** [comment]
All tests pass locally.  Noting that `GainMixin` is not tested explicitly:

```py
$ git grep GainMixin
apstools/devices/srs570_preamplifier.py:class GainMixin:
apstools/devices/srs570_preamplifier.py:class GainSignal(GainMixin, EpicsSignal):
```

File: apstools/devices/srs570_preamplifier.py (hunk 2 above)
Comment (canismarko) [reply apstools/devices/srs570_preamplifier.py:178]: It was originally going to be used in a pcdsdevices MultiDerivedSignal class. I re-factored it to no longer need it that way. Happy to re-combine the mixin with the signal.

Comment (prjemian) [reply apstools/devices/srs570_preamplifier.py:178]: Better to recombine if you do not see an additional use.

File: apstools/devices/srs570_preamplifier.py [hunk 3]
```diff
@@ -21,12 +21,169 @@
...
+    # Get calibrated settle time, or None to use the Ophyd default
+    return settling_times.get((gain_value, gain_unit, gain_mode))
+
+
+class GainMixin:
+    """A signal where the settling time depends on the pre-amp gain.
+
+    Used to introduce a specific settle time when setting to account
+    for the amp's R–C relaxation time when changing gain.
+
```
Comment (prjemian) [inline apstools/devices/srs570_preamplifier.py:122]: this blank line can be deleted

File: apstools/devices/srs570_preamplifier.py [hunk 4]
```diff
@@ -21,12 +21,169 @@
...
+        Returns
+        -------
+        st : Status
+
+        See Also
+        --------
+        Signal.set
+        EpicsSignal.set
+
```
Comment (prjemian) [inline apstools/devices/srs570_preamplifier.py:154]: Is this blank line needed?

File: apstools/devices/srs570_preamplifier.py [hunk 5]
```diff
@@ -21,12 +21,169 @@
...
+            _settle_time = settle_time
+        return super().set(value, timeout=timeout, settle_time=_settle_time)
+
+
+class GainSignal(GainMixin, EpicsSignal):
+    """A signal where the settling time depends on the pre-amp gain.
+
+    Used to introduce a specific settle time when setting to account
+    for the amp's R–C relaxation time when changing gain.
+
```
Comment (prjemian) [inline apstools/devices/srs570_preamplifier.py:183]: Is this blank line needed?

**Review by prjemian — APPROVED** [review]
LGTM!  @canismarko - Please press the big green Merge button.

**Pull request merged**